broadinstitute/single_cell_portal_core
# Enable server-side profiling for any page (SCP-3797)

This adds a configuration setting for profiling server-side code with `rack-mini-profiler`, to help optimization engineering.  Setting `config.profile_performance = true` will add a small "speed badge" at top left on any page.  Clicking that opens more data to explore.  

This complements existing profiling via SCP's `GenericProfiler` ([PR](https://github.com/broadinstitute/single_cell_portal_core/pull/969), [docs](https://docs.google.com/document/d/1oSJklX9bsO4AGEwHwvZurZqnVTd6kWNz4OiZ4Zu2GtE/edit#)).  Whereas `GenericProfiler` is great for investigating timing data for a specific method, [`rack-mini-profiler`](https://github.com/MiniProfiler/rack-mini-profiler#rack-mini-profiler) is great for investigating granular timing data for a general page -- often comprised of functions that are not otherwise feasible to profile.

Here's how this profiling looks when enabled:

https://user-images.githubusercontent.com/1334561/139253987-f723f5ac-5e86-4ea0-98b0-d39f555855a5.mov

Adding `?pp=flamegraph` to the URL will show a flamegraph, and `?pp=help` will show how to do even more.  More details and particular findings are [documented here](https://docs.google.com/presentation/d/1rvenYVO6mODe2thUDmWvFCveA5xh7lpE218i_XAixm0/edit).

To test:
* Pull this branch, `bundle install`, `rails s`
* Go to a page, see no profiling button at top left of page
* In `config/environments/development.rb`, set `config.profile_performance = true`
* Restart `rails`
* Go to a page, see profiling button at top left of page.  Click on it, verify more detail is shown.
* In `config/environments/development.rb`, set `config.profile_performance = false`

This relates to SCP-3797.
